### PR TITLE
tests: run integration test on macos GH runners as well

### PR DESCRIPTION
**Pull request by mvo5**
This is a rebase/update of https://github.com/osbuild/bootc-image-builder/pull/33 without hacks.

### CODE REVIEW

**Comment by achilleas-k** [comment]
Instead of relying on the journal for logs and losing the checks on macOS, we could run the `podman run` command without `--rm` and then fetch the logs with `podman logs <container>`.
Worth it?

**Comment by mvo5** [comment]
> Instead of relying on the journal for logs and losing the checks on macOS, we could run the `podman run` command without `--rm` and then fetch the logs with `podman logs <container>`. Worth it?

Love the suggestion - added now!

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -76,3 +76,25 @@ jobs:
...
         sudo -E XDG_RUNTIME_DIR= pytest-3 -s -vv
+
+  integration-macos:
+    name: "Integration macos"
+    # needed to get latest cpu
+    runs-on: macos-13
+    steps:
+    - uses: actions/checkout@v4
+      with:
+        ref: ${{ github.event.pull_request.head.sha }}
```
Comment (lmilbaum) [inline .github/workflows/tests.yml:87]: Is the checkout action uses this ref by default? Is that addition necessary?

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -76,3 +76,25 @@ jobs:
         # XDG_RUNTIME_DIR is set.
         # TODO: figure out what exactly podman needs
         sudo -E XDG_RUNTIME_DIR= pytest-3 -s -vv
+
+  integration-macos:
+    name: "Integration macos"
+    # needed to get latest cpu
+    runs-on: macos-13
+    steps:
+    - uses: actions/checkout@v4
```
Comment (lmilbaum) [inline .github/workflows/tests.yml:85]: I'd suggest using the full action version (vX.Y.Z) because there might be rare cases where action versions are not backward compatible.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -76,3 +76,25 @@ jobs:
...
+    - name: Setup up python
+      uses: actions/setup-python@v4
+    - name: Setup up podman
+      run: |
+        brew install podman netcat
+        podman machine init --rootful
+        podman machine start
+    - name: Install test dependencies
+      run: |
+        sudo pip install pytest flake8 paramiko
```
Comment (lmilbaum) [inline .github/workflows/tests.yml:97]: Would you mind using a requirements.txt file to list the packages with their corresponding pinned versions?

File: .github/workflows/tests.yml (hunk 1 above)
Comment (mvo5) [reply .github/workflows/tests.yml:87]: Thank you, that is a nice catch. TBH I just cargo-culted this from the rest of the file. I will clean it up everywhere :)

Comment (mvo5) [reply .github/workflows/tests.yml:87]: Actually, I looked at this more closely and there is a subtle difference. The default behavior is AFAICT that a merge commit of main and the PR is generated and checked out. With our current setup we only get the PR HEAD [1].

I have no strong opinion either way but would prefer to consult with the team first (as I'm relatively new and lack a lot of the history) what the rational for this decision is and if it needs changing I will do it in a separate PR. Hope that is okay with you :)

[1] https://github.com/actions/checkout#checkout-pull-request-head-commit-instead-of-merge-commit

File: .github/workflows/tests.yml (hunk 2 above)
Comment (mvo5) [reply .github/workflows/tests.yml:85]: Thanks for this suggestion. I would like to get input from @ondrejbudai or @achilleas-k because AFAICT we are not doing this in any of our other repos. I'm happy to check this but want it to be consistent (in this repo at least). It will be a separate PR that changes it for all checkout actions in the workflow. Hope that is okay?

File: .github/workflows/tests.yml (hunk 3 above)
Comment (mvo5) [reply .github/workflows/tests.yml:97]: Thanks, good point. Let me fix this.

File: .github/workflows/tests.yml (hunk 2 above)
Comment (ondrejbudai) [reply .github/workflows/tests.yml:85]: No strong opinion here, I generally like pinning deps, but I have never seen an issue with `actions/checkout` (and our use-case is trivial), so I don't care if this particular one is pinned.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (ondrejbudai) [reply .github/workflows/tests.yml:87]: I think that one advantage is of using a PR head is that you clearly know what commit is being tested. But it's matter of taste, tbh.

I'm more interested if this works correctly when run on a merge queue, it looks a bit fishy to me tbh.

**Comment by ondrejbudai** [comment]
This test takes 50 minutes to run on the Github-hosted worker. Should we try deploying a self-hosted runner on our AWS machine?

File: .github/workflows/tests.yml (hunk 1 above)
Comment (achilleas-k) [reply .github/workflows/tests.yml:87]: That's a good point, because on the merge queue, the `github.event.pull_request` should be empty, so I guess `ref` is unset then.
Comparing https://github.com/osbuild/bootc-image-builder/actions/runs/7384902395/job/20088639130#step:2:2 (checkout action for shellcheck on a PR) with https://github.com/osbuild/bootc-image-builder/actions/runs/7386384027/job/20092854868#step:2:2 (checkout action for shellcheck on the merge queue) seems to support this.

Comment (mvo5) [reply .github/workflows/tests.yml:87]: If I read the room correctly here it sounds like we do want to move to testing the merge commit instead (at least on systems that use a merge queue)? Fwiw, I tend to agree but have no strong opinion either way. I do care that we do it consistently though :) Unless someone here disagrees I will prepare PRs to make this change consistently.

**Comment by mvo5** [comment]
> This test takes 50 minutes to run on the Github-hosted worker. Should we try deploying a self-hosted runner on our AWS machine?

Great point  - I setup our AWS machine as a self hosted runner and it finisehd in 14min. so definitely worth it. Self hosting on macOS in AWS is a bit of a PITA though because of https://github.com/actions/runner/issues/947 and https://github.com/actions/runner/issues/1056 - I will see what I can do about that.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (mvo5) [reply .github/workflows/tests.yml:89]: I will do this in a followup if that's okay? Would like to not block on it for this PR.

**Comment by mvo5** [comment]
> > This test takes 50 minutes to run on the Github-hosted worker. Should we try deploying a self-hosted runner on our AWS machine?
> 
> Great point - I setup our AWS machine as a self hosted runner and it finisehd in 14min. so definitely worth it. Self hosting on macOS in AWS is a bit of a PITA though because of [actions/runner#947](https://github.com/actions/runner/issues/947) and [actions/runner#1056](https://github.com/actions/runner/issues/1056) - I will see what I can do about that.

This is ready now, sorry that it took a bit, the self hosted follows the setup from https://github.com/actions/runner/issues/1056#issuecomment-1237426462 and is now running a macOS background service.

**Review by achilleas-k — APPROVED** [review]

**Review by ondrejbudai — APPROVED** [review]
:apple: Go!
